### PR TITLE
HPCC-27671 Create confidence test and key file for the General Linear Model Bundle

### PR DESCRIPTION
**Pull request by NSel1727**
• Create a directory titled ecl which will contain test files and a key folder
• Make a new file called "FamilyTest" that takes code from the BinomialRegression file from the validation folder and makes it functional so each linear model family can be tested
• Implement version control for the FamilyTest 
• Create a key file

### CODE REVIEW

**Comment by NSel1727** [comment]
@AttilaVamos

File: ecl/FamilyTest.ecl [hunk 1]
```diff
@@ -0,0 +1,177 @@
+/*##############################################################################
+## HPCC SYSTEMS software Copyright (C) 2022 HPCC Systems®.  All rights reserved.
+############################################################################## */
+
+// version tmod='Binomial'
```
Comment (AttilaVamos) [inline ecl/FamilyTest.ecl]: There should be no space between '//' and 'version'

File: ecl/FamilyTest.ecl [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+myMod := #IFDEFINED(root.tmod, 'Full');
+
+#IF(myMod = 'Binomial')
+	fam := Family.Binomial;
+	vals := DATASET([{'sm.logit',
+	7.32292705, -0.25274345, -2.77938918,  1.29930595, -2.70427087,
+	2.49799533,  0.64945102,  0.78587716,  0.68228159,  1.16265393,
+	0.00337306,  0.69715427,  0.00040520,  0.05686405,  0.02002140,
+	155.76486509, 1}], Test_Values);
+#ElseIf(myMod = 'Quasibinomial')
```
Comment (AttilaVamos) [inline ecl/FamilyTest.ecl]: Please use all capitals for '#ELSEIF' (it is coding style what we recommend).

File: ecl/key/FamilyTest.xml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<Dataset name='TestName'>
+ <Row><TestName>GLM -- FamilyTest</TestName></Row>
+</Dataset>
+<Dataset name='Result'>
+ <Row><Result>Pass</Result></Row>
+</Dataset>
+<Dataset name='Errors'>
+</Dataset>
+
```
Comment (AttilaVamos) [inline ecl/key/FamilyTest.xml]: Please remove this empty line from the key file.

File: ecl/key/FamilyTest.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<Dataset name='TestName'>
+ <Row><TestName>GLM -- FamilyTest</TestName></Row>
+</Dataset>
+<Dataset name='Result'>
+ <Row><Result>Pass</Result></Row>
+</Dataset>
+<Dataset name='Errors'>
+</Dataset>
```
Comment (AttilaVamos) [inline ecl/key/FamilyTest.xml]: Okay, a single '\n' (0x0A) char needed at the end of that line. I know Windows uses '\r\n' by default.

**Comment by NSel1727** [comment]
@AttilaVamos @RogerDev @ghalliday 

Please review my pull request. 

All of the seven test cases obtain the correct output but two of them go through an exception. Here is the JIRA issue: https://track.hpccsystems.com/browse/ML-513

File: ecl/FamilyTest.ecl [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+		IrisData enum_recs(Datasets.Iris rec, UNSIGNED c) := TRANSFORM
+			SELF.id := c;
+			SELF.class := IF(rec.class=1, 1, classRange);
+			SELF := rec;
+		END;
+		
+		iris := PROJECT(Datasets.Iris, enum_recs(LEFT,COUNTER));
+
+		Core.ToField(iris, iris_indep, id, , 1,
+								 'sepal_length,sepal_width,petal_length,petal_width');
```
Comment (AttilaVamos) [inline ecl/FamilyTest.ecl]: Indentation problem.

File: ecl/FamilyTest.ecl [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+			REAL8 std_value;
+			REAL8 tst_value;
+			BOOLEAN equal;
+		END;
+
+		Compare_Rec check(GLM_Types.pval_Model_Coef p, Test_Values t, UNSIGNED s):=TRANSFORM
+			SELF.src := t.src;
+			SELF.tst_value := CHOOSE(s, p.w, p.se, p.p_value);
+			SELF.std_value := CHOOSE(p.ind_col+1,
+															 CHOOSE(s, t.icept_coef, t.icept_err, t.icept_pval),
```
Comment (AttilaVamos) [inline ecl/FamilyTest.ecl]: Tab vs spaces (?) indentation problems.

File: ecl/FamilyTest.ecl [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	fam := Family.InvGauss;
+	vals := DATASET([{'sm.logit',
+			0.8292705, 0.00164345, -0.177038918,  0.0909830595, -0.213827087,
+			0.18899533,  0.054845102,  0.054387716,  0.052728159,  0.086865393,
+			0.00005937306,  0.97515427,  0.00040520,  0.0845286405,  0.01326702140,
+			300.28486509, 1}], Test_Values);
+	classRange := 2;
+#END
+
+errors := FamilyTest(fam, vals, classRange);
```
Comment (AttilaVamos) [inline ecl/FamilyTest.ecl:191]: Nice use of 'classRange' variable to fix that exception problem.

**Review by AttilaVamos — CHANGES_REQUESTED** [review]
Seems good, but there are some small issues.

**Comment by NSel1727** [comment]
@AttilaVamos

**Review by AttilaVamos — COMMENTED** [review]
I think that 8 spaces indentation is a bit wide and sometimes it makes the review hard. Usually, we use 4 spaces.  

I would split the SEQUENTIAL(...) into multiline to avoid too lengthy line. 

Please ask the others for review.

**Comment by ghalliday** [comment]
> I think that 8 spaces indentation is a bit wide and sometimes it makes the review hard. Usually, we use 4 spaces.
> 
> I would split the SEQUENTIAL(...) into multiline to avoid too lengthy line.
> 
> Please ask the others for review.

I think the problem is that the file contains tabs rather than spaces - and the local tab setting is different from the 8 used by github.  The general policy is to only uses spaces for indentation to avoid the problem.

**Comment by NSel1727** [comment]
@ghalliday @RogerDev 

Please review this pull request for if it is ready to be merged.

**Review by AttilaVamos — APPROVED** [review]
Looks good from automated testing point of view.

**Comment by ghalliday** [comment]
I had a couple of questions - use of tabs rather than spaces, and whether the copyright message is in the correct format.
@RogerDev are you happy with this - if so I'll merge.

**Comment by NSel1727** [comment]
I will update the copyright message.

**Pull request merged**